### ISSUE TITLE
Unable to type # and { characters in Windows version

### ISSUE THREAD

**Issue by flemmens** [mapbox/mapbox-studio-classic#878]
All is in the title, it is impossible to type # or { characters in the style sheet, the only way I have found is to use copy/paste.

**Comment by yhahn** [comment]
Hrm, I'm unable to reproduce this. Anything special about your setup (keyboard/otherwise?) that could help us track this down?

**Comment by wilhelmberg** [comment]
Strange, I can partially reproduce with MBS 0.1.2 on W7, W8.1 and Server 2012.
All with GUI set to EN and keyboard set to DE:

`#` works
`}` works
`{` doesn't work

**Comment by flemmens** [comment]
Nothing special about my keyboard, this is the only software where I have this issue. It works correctly in Tilemill too. I use a French/Belgian keyboard.

**Comment by wilhelmberg** [comment]
`[` and `]` also don't work for me.

**Comment by springmeyer** [comment]
Can anyone replicate this in the new v0.2.0 release? (https://www.mapbox.com/mapbox-studio). I'm hoping it was an atom shell bug. If it still persists can you check if the problem also exists if you load up Mapbox Studio at http://localhost:3000 in your web browser?

**Comment by springmeyer** [comment]
closing unless/util this is confirmed again.

**Comment by wilhelmberg** [comment]
These still don't work for me: `{`, `[`, `]`
Win7, Win8.1

**Comment by springmeyer** [comment]
okay :)

**Comment by wilhelmberg** [comment]
Forgot: doesn't make a difference if I start `mapbox.exe` (atom-shell) or `node index.js` (browser, tested with latest IE, FF, Chrome).

**Comment by wilhelmberg** [comment]
Using [ALT codes](http://usefulshortcuts.com/alt-codes/punctuation-alt-codes.php) I can enter `{` via holding down `Alt`-key and typing `123` on the numeric keyboard.

**Comment by wilhelmberg** [comment]
Comparing the keyboard events from the demo on the startpage of http://codemirror.net/ (which accepts `{`) and Studio the only difference I can see is the `returnValue` property.
On codemirror.net it's `true` and in Studio it's `false`.

KeyboardEvent in Studio (Germany keyboard layout) for `{`:

![keyboard-event](https://cloud.githubusercontent.com/assets/4549888/5180612/b8a97f7c-748d-11e4-8483-ec1486aa4d73.png)

**Comment by yhahn** [comment]
O_O

**Comment by gendronb** [comment]
Just installed the latest version yesterday, they don't work for me either : {, [, ], }

Windows 7 Pro...

**Comment by amyleew** [comment]
Windows 7 SP1 ... same error

**Comment by wilhelmberg** [comment]
@flemmens @gendronb 
Do you have `English US` as keyboard language in your Windows system settings?
Or another language?
I'm able to type the missing characters when I switch keyboard language to `English US`

---

As the codemirror samples on the web all work with my German language setting, I suppose the characters get swallowed somewhere in the `carto.css` definition.
The key codes are different for different for different keyboard langugaes.
I've dug through Mapbox Studio and codemirror source, but wasn't yet able to find where this happens.

**English US** keyboard layout, key code for **`{`**:
![image](https://cloud.githubusercontent.com/assets/4549888/6741257/50275250-ce86-11e4-9bbb-1a9ff5c18fa1.png)

**German AT** keyboard layout, key code for **`{`**:
![image](https://cloud.githubusercontent.com/assets/4549888/6741264/5a76350a-ce86-11e4-81d0-080f0cc2febc.png)

**Comment by stephankn** [comment]
I had an old version of Mapbox Studio and never noticed this problem. After being absent for a while I recently updated to 0.2.7. Now I can't type "{" or "[" in the style editor any more.

Mapbox Studio 0.2.7.
Windows 8.1 64bit
English US System language
Region German
German (IBM) keyboard

input on codemirror.net works fine.

Quite sad as this renders Mapbox Studio practically useless.

**Comment by wilhelmberg** [comment]
@stephankn There's a workaround, if you have a keyboard with a numeric keypad, you can use `Alt` codes to create the missing braces:
- Press (and keep pressed) `Alt`
- Type numeric code of character you want
- Release `Alt`

`Alt` + `123` => `{`
`Alt` + `125` => `}`
`Alt` + `91` => `[`
`Alt` + `93` => `]`

---

Progress report:
- I concentrated on `}` (works) and `{` (doesn't work)
- same problem exists with codemirror 5.3.0 (released yesterday)
- somewhere/somehow the `keypress` event gets mangled/catched before codemirror can properly process it
- even setting Chrome's `Event Listener Breakpoints` just gets`keyup` for `{` instead of `keypress` and `input` events, like for `}`
- I removed every reference to `carto` mode and definitions, so that's ruled out as a cause
- `{` works with the online codemirror demos

---

**I created a minimum testcase:**

``` html
<html>
<head>
<script src="../ext/codemirror/codemirror.js"></script>
<link rel="stylesheet" href="../ext/codemirror/codemirror.css">
</head>
<body>
<script type="text/javascript">
    var myCodeMirror = CodeMirror(document.body);
</script>
</body>
```

to investigate KeyboardEvents.

**Event for `}` (works) looks the same in minimum demo and in Mapbox Studio:**

![image](https://cloud.githubusercontent.com/assets/4549888/7748976/cf2fcfcc-ffc9-11e4-821c-c3666d3f5d1c.png)

**Event for `{` in minimum demo:**

![image](https://cloud.githubusercontent.com/assets/4549888/7749123/0ea52a48-ffcb-11e4-8443-30ebaf1d28ef.png)

**Event for `{` in Mapbox Studio:**

![image](https://cloud.githubusercontent.com/assets/4549888/7749180/84764b62-ffcb-11e4-8b5a-fac7c52f4e79.png)

Differences of events:
- charCode: `0` vs. `123`
- keyCode: `55` vs. `123`
- keyIdentifier: `U+0037` vs. `F12`
- type: `keyup` vs `keypress`
- which: `55` vs. `123`

**Comment by wilhelmberg** [comment]
@flemmens @yhahn @springmeyer @gendronb @amyleew @stephankn 

**FINALLY!!!**
After spending endless days (and some nights) on stepping through MapBox Studio and codemirror I found it.

The events for `{` and `[` get swallowed by the shortcuts to switch between tabs.

---

That's what the sequence of events looks on my keyboard, when I press `{` (AltGr+7).

![image](https://cloud.githubusercontent.com/assets/4549888/8061989/b96fcf80-0ecd-11e5-8aae-fa65a694bbbe.png)

Because of `keyCode 55` falling into the range `(which > 48 && which < 58)` `Editor.prototype.keys` returns `false` and thus property `preventDefault` of the `keydown` event gets set to `true` and thus the `keypress` event for `{` never fires.

``` javascript
  case ((which > 48 && which < 58) && ev.altKey): // 1-9
    var tab = $('#tabs a.tab')[(which-48)-1];
    if (tab) $(tab).click();
    break;

```

---

### Quick fix (for an existing install)

Remove the above case statement (or adjust `(which > 48 && which < 58)` to not lose tab shortcuts), which can be found here:

https://github.com/mapbox/mapbox-studio/blob/c5c21df2b3442166bb7f82a68a60c86bac5f576e/app/style.js#L345-L348

**Comment by springmeyer** [comment]
Incredible work @BergWerkGIS!!!

**Comment by wilhelmberg** [comment]
Thanks :rocket:

**Comment by amyleew** [comment]
Nice work @BergWerkGIS!!!

**Comment by wilhelmberg** [comment]
@amyleew thanks, :smiley: :smiley: :smiley: 

Do you use tab shortcuts?
How many tabs do you have in an average project?

Do you think, people will have more than 6 and want to switch between them?

**Comment by stephankn** [comment]
I never used tab shortcuts. Was not aware of them. Don't need them.

Avoiding hard-coded key-codes sounds like a good idea in general as long as you don't know what keyboard layout they might conflict with.

Is the following a possible solution?
http://www.w3.org/TR/DOM-Level-3-Events/#keys

**Comment by wilhelmberg** [comment]
@stephankn thanks, UI Events (formerly DOM Level 3 Events) might help. We will take a look.

Have you tried my suggested solution?
Does it work for you now?

**Comment by stephankn** [comment]
I can confirm that after commenting out the last case block in style.js as you suggested keys work.

I just looked around that file and found some codes I thought they might be layout dependent.
Tried Ctrl + / which should open help and Ctrl + \ which is supposed to open settings,

Both didn't work. No big issue for me, but likely same root cause.

On German layout for "/" you press Shift+7. So That hot-key would be Ctrl+Shift+7
On German layout for "\" you press AltGr+"ß" (same key that would produce with Shift a "?"). So that key would be Ctrl+AltGr+ß

No idea what key would produce code 191 or 220.
